UniversalAlgebra/Complexity/FiniteCSP.lagda

```diff
@@ -0,0 +1,90 @@
+---
+layout: default
+title : Complexity.FiniteCSP module (The Agda Universal Algebra Library)
+date : 2021-07-14
+author: [agda-algebras development team][]
+---
+
+### Constraint Satisfaction Problems
+
+\begin{code}
+
+{-# OPTIONS --without-K --exact-split --safe #-}
+
+module Complexity.FiniteCSP where
+
+open import Agda.Primitive         using    ( _⊔_ ; lsuc ; Level)
+                                   renaming ( Set to Type )
+open import Data.Product           using    ( _,_ ; Σ ; Σ-syntax )
+open import Data.Nat.Base          using    ( ℕ )
+open import Data.Fin.Base          using    ( Fin )
+open import Function.Base          using    ( _∘_ )
+open import Relation.Unary         using    ( _∈_; Pred   )
+
+open import Relations.Continuous   using    ( Rel )
+open import Algebras.Basic         using    ( Signature )
+
+\end{code}
+
+
+#### Finite CSP
+
+An instance of a (finite) constraint satisfaction problem is a triple 𝑃 = (𝑉, 𝐷, 𝐶) where
+
+* 𝑉 is a finite set of variables,
+* 𝐷 is a finite domain,
+* 𝐶 is a finite list of constraints, where
+  each constraint is a pair 𝐶 = (𝑥, 𝑅) in which
+  * 𝑥 is a tuple of variables of length 𝑛, called the scope of 𝐶, and
+  * 𝑅 is an 𝑛-ary relation on 𝐷, called the constraint relation of 𝐶.
+
+\begin{code}
+
+private variable χ ρ : Level
+
+record Constraint (∣V∣ ∣D∣ : ℕ) : Type (lsuc (χ ⊔ ρ)) where
+ field
+  scope : Fin ∣V∣ → Type χ -- a subset of Fin ∣V∣
```

Comment:
This seems overly general. There is little point in having witnesses to whether something is in scope.  You might as well use `Bool` instead. And then `Vec (Fin |V|) Bool` might be easier?

Reply:
Because `Fin |V|` has decidable equality (and is finite!), you gain very little from having `scope` be so very general. You may as well have other things also have decidable equality.